keel-ec2/src/main/kotlin/com/netflix/spinnaker/keel/ec2/asset/AmazonSecurityGroupHandler.kt

```diff
@@ -124,18 +130,67 @@ class AmazonSecurityGroupHandler(
       .apply {
         typeMetadata = request.asset.typeMetadata
         spec = SecurityGroupProto.newBuilder()
-          .also {
-            it.name = name
-            it.accountName = accountName
-            it.region = region
-            it.vpcName = vpcId?.let { cloudDriverCache.networkBy(it).name }
-            it.description = description
+          .also { builder ->
+            builder.name = name
+            builder.accountName = accountName
+            builder.region = region
+            builder.vpcName = vpcId?.let { cloudDriverCache.networkBy(it).name }
+            builder.description = description
+            inboundRules.forEach { rule ->
+              builder.inboundRuleList.add(
+                SecurityGroupRule.newBuilder().also { builder ->
+                  val referencedGroup = rule.securityGroup
+                  val cidrRange = rule.range
+                  when {
+                    referencedGroup != null && (referencedGroup.vpcId == vpcId && referencedGroup.name == name) -> {
+                      builder.selfReferencingRuleBuilder.also { ruleBuilder ->
+                        ruleBuilder.protocol = rule.protocol
+                        rule.portRanges?.forEach {
+                          ruleBuilder.portRangeList.add(it.toProto())
+                        }
+                      }
+                    }
+                    referencedGroup != null && (referencedGroup.accountName != accountName || referencedGroup.region != region) -> {
```

Comment:
If account or region differ from the parent security group, it's a cross-region rule.